src/ripple/app/tx/impl/DID.cpp

```diff
@@ -0,0 +1,234 @@
+        if (auto field = ctx.tx[~sField])
+            return field->length() > length;
+        return false;
+    };
+
+    if (isTooLong(sfURI, maxDIDURILength) ||
+        isTooLong(sfDIDDocument, maxDIDDocumentLength) ||
+        isTooLong(sfData, maxDIDAttestationLength))
+        return temMALFORMED;
+
+    return preflight2(ctx);
+}
+
+TER
+DIDSet::preclaim(PreclaimContext const& ctx)
+{
+    if (!ctx.view.read(keylet::did(ctx.tx[sfAccount])) &&
+        !ctx.tx.isFieldPresent(sfURI) && !ctx.tx.isFieldPresent(sfDIDDocument))
+    {
+        // Need either the URI or document if the account doesn't already have a
+        // DID
+        return tecEMPTY_DID;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+addSLE(
+    ApplyContext& ctx,
+    std::shared_ptr<SLE> const& sle,
+    AccountID const& owner)
+{
+    auto const sleAccount = ctx.view().peek(keylet::account(owner));
+    if (!sleAccount)
+        return tefINTERNAL;
+
+    // Check reserve availability for new object creation
+    {
+        auto const balance = STAmount((*sleAccount)[sfBalance]).xrp();
+        auto const reserve =
+            ctx.view().fees().accountReserve((*sleAccount)[sfOwnerCount] + 1);
+
+        if (balance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    // Add ledger object to ledger
+    ctx.view().insert(sle);
+
+    // Add ledger object to owner's page
+    {
+        auto page = ctx.view().dirInsert(
+            keylet::ownerDir(owner), sle->key(), describeOwnerDir(owner));
+        if (!page)
+            return tecDIR_FULL;
+        (*sle)[sfOwnerNode] = *page;
+    }
+    adjustOwnerCount(ctx.view(), sleAccount, 1, ctx.journal);
+    ctx.view().update(sleAccount);
+
+    return tesSUCCESS;
+}
+
+TER
+DIDSet::doApply()
+{
+    // Edit ledger object if it already exists
+    Keylet const didKeylet = keylet::did(account_);
+    if (auto const sleDID = ctx_.view().peek(didKeylet))
+    {
+        auto update = [&](auto const& sField) {
+            if (auto const field = ctx_.tx[~sField])
+            {
+                if (field->empty())
+                {
+                    sleDID->makeFieldAbsent(sField);
+                }
+                else
+                {
+                    (*sleDID)[sField] = *field;
+                }
+            }
+        };
+        update(sfURI);
+        update(sfDIDDocument);
+        update(sfData);
+
+        if (!sleDID->isFieldPresent(sfURI) &&
+            !sleDID->isFieldPresent(sfDIDDocument))
+        {
+            return tecEMPTY_DID;
+        }
+        ctx_.view().update(sleDID);
+        return tesSUCCESS;
+    }
+
+    // Create new ledger object otherwise
+    auto const sleDID = std::make_shared<SLE>(didKeylet);
+    (*sleDID)[sfAccount] = account_;
+
+    auto set = [&](auto const& sField) {
+        if (auto const field = ctx_.tx[~sField];
+            field.has_value() && !field->empty())
+            (*sleDID)[sField] = field.value();
```

Comment:
We tend not to use `has_value` and `value` . Consider rewriting this in a similar style to your `update` lambda above. I.e. something like:
```
        if (auto const field = ctx_.tx[~sField]; field && !field->empty())
            (*sleDID)[sField] = *field;
```

Reply:
Fixed in https://github.com/XRPLF/rippled/pull/4636/commits/6bda30b852d448b4b3df2d534b6441ed0af92663